### PR TITLE
Parse uppercase SRS files

### PR DESCRIPTION
**Pull request by PaulJWright**
This MR modifies `sunpy.io.special.srs` to work with uppercase files. This reduces the number of load errors in the dataset (~9800 files from 1996 - present) from 17% down to 0.3%.

The number of upper-case and mixed-case files in the dataset:

  | File-type       | Number of Occurrences |
  |-----------------|----------------------|
  | Upper-case      |         1681         |
  | Mixed-case      |         8109         |

This is accomplished by:

1. **Modifying Column Names**: The column names in `make_table()` are cast as title case using the `.title()` method. This ensures consistent formatting and readability across uppercase and mixed-case SRS files.
2. **Using Regex**: Regular expressions are used to perform case-insensitive replacements or matching. For example, one use-case is to remove the space (and format as title case) in the column name "Mag Type" (regardless of case) and replace with "Magtype" in `split_lines()` e.g.: replacing the first line with second:

      ```python
      t1_lines[1] = t1_lines[1].replace('Mag Type', 'MagType')  # original that does not work with uppercase files
      ```
      ```python
      t1_lines[1] = re.sub(r'Mag\s*Type', r'Magtype', t1_lines[1], flags=re.IGNORECASE)  # case-insensitive replacement
      ```
3. **Parsing Supplementary Sections**: `split_lines()` is modified to handle supplementary sections that may occur after the main data sections (`I.`, `IA`, `II`). The lines starting with `III`, `COMMENT`, `EFFECTIVE 2 OCT 2000`, `PLAIN`, `This message is for users of the NOAA/SEC Space` and `NNN` are identified and if `>= 1` occurrence is found, the first line is considered as the final section header (as long as it occurs on a line after the main section headers). 

      | Lines                   | Number of Occurrences | Included |
      |--------------------------|----------------------|----------------------|
      | `III `                     | 18                     | Y |
      | `COMMENT`                  | 17                     | Y |
      | `EFFECTIVE 2 OCT 2000`     | 19                     | Y |
      | `PLAIN`                    | 19                     | Y |
      | `This message is for users of the NOAA/SEC Space` | 18                     | Y |
      | `NNN`                      | 23                     | Y |
      | `99999`                      | 5                     | N |

      This is necessary as in the current implementation the last section (`IA`) may contain additional lines of the file that aren't rows of the table (this raises an error during parsing). An example SRS file with a final section starting with `III` is shown below:
    ```txt
    :Product: 20000926SRS.txt
    :Issued: 2000 Sep 26 0030 UTC
    # Prepared jointly by the U.S. Dept. of Commerce, NOAA,
    # Space Environment Center and the U.S. Air Force.
    #
    JOINT USAF/NOAA SOLAR REGION SUMMARY
    SRS NUMBER 270 ISSUED AT 0030Z ON 26 SEP 2000
    REPORT COMPILED FROM DATA RECEIVED AT SWO ON 25 SEP
    I.  REGIONS WITH SUNSPOTS.  LOCATIONS VALID AT 25/2400Z
    NMBR LOCATION  LO  AREA  Z   LL   NN MAG TYPE
    9166 S12W70   125  0380 DAI  10   23 BETA
    9167 N15W42   097  0080 CSO  06   05 BETA
    9169 N11W22   077  1450 FKC  20   88 BETA-GAMMA-DELTA
    9170 S06W27   082  0080 DAO  10   19 BETA
    9171 S15E33   022  0080 CSO  07   03 BETA
    9172 N12E58   357  0100 DAO  06   06 BETA
    9173 S13E60   355  0070 HAX  03   01 ALPHA
    IA. H-ALPHA PLAGES WITHOUT SPOTS.  LOCATIONS VALID AT 25/2400Z SEP
    NMBR  LOCATION  LO
    NONE
    II. REGIONS DUE TO RETURN 26 SEP TO 28 SEP
    NMBR LAT    LO
    NONE
    III. COMMENTS
    EFFECTIVE 2 OCT 2000, SEC WILL ISSUE ITS TEXT PRODUCTS IN MIXED
    CASE VS. ALL UPPER CASE.  THIS CHANGE WILL BE VISIBLE ON ALL SEC
    PRODUCT DELIVERY SYSTEMS EXCEPT FOR TELETYPE CIRCUITS.  ONLY THE
    CASE OF THE TEXT IN EACH PRODUCT WILL CHANGE, NOT THE FORMAT OR
    SPACING WITHIN THE PRODUCT.  FOR DETAILS SEE OUR WEBSITE AT
    SEC.NOAA.GOV/MIXEDCASE.HTML
    ```
4. **Handling Empty Column `COMMENT`**: For SRS files before 2000-10-02, there is often (326 occurrences) an empty column `COMMENT` in section `IA` (`t2_lines`). If the `IA` section contains a `COMMENT` column, the new function `_try_drop_column()` tries to drop the column (if empty, and the values in the rows are consistent for what is expected for other columns) by comparing to an `expected_pattern_dict`:
      ```python
      expected_pattern_dict = {
          'Nmbr': r'^\d+$',
          'Location': r'[NESW]\d{2}(?:[NESW]\d{2})?',  # can be e.g. S02E30 or S02, etc...
          'Lo': r'^\d+$',
          }
      ```
     where an example of an SRS file with this empty `COMMENT` column is shown below:
    ```txt
    :Product: 19960301SRS.txt
    :Issued: 1996 Mar 01 0030 UTC
    # Prepared jointly by the U.S. Dept. of Commerce, NOAA,
    # Space Environment Center and the U.S. Air Force.
    #
    JOINT USAF/NOAA SOLAR REGION SUMMARY
    SRS NUMBER 061 ISSUED AT 0030Z ON 01 MARCH 1996
    REPORT COMPILED FROM DATA RECEIVED AT SESC ON 29 FEBRUARY
    I. REGIONS WITH SUNSPOTS. LOCATIONS VALID AT 29/2400Z FEBRUARY
    NMBR LOCATION  LO  AREA  Z   LL   NN MAG TYPE
    7946  N07W53  195  0000 AXX  00  001 ALPHA
    7949  S12W07  149  0020 BXO  04  006 BETA
    IA. H-ALPHA PLAGES WITHOUT SPOTS.  LOCATIONS VALID AT 29/2400Z
    NMBR  LOCATION  LO            COMMENT
    7948  N10W96   238
    II. REGIONS DUE TO RETURN 01 MARCH TO 03 MARCH
    NMBR LAT    LO
    NONE
    ```

closes: https://github.com/sunpy/sunpy/issues/7034

### To-do

- [x] Unit tests.
- [x] Investigate other potential common loading issues in pre-`2000-10-03` files
- [x] add `CHANGELOG` based final MR scope

### Questions

- [x] Do we want to cast McIntosh/Hale classifications to titlecase?
- [x] Add sections after `IA` to metadata?
- [x] While this MR reduces the number of errors down from 17% to ~0.3%, there is very infrequently, a spurious `99999` at the end of the SRS file (shown below). Using this in the `split_lines()` seems safe enough, but there are very few occurrences and I know that after 2002-06-15, reformatting of numbers needs to occur (``out_table['Number'] += 10000``). :
    ```txt
    # Prepared jointly by the U.S. Dept. of Commerce, NOAA,
    # Space Environment Center and the U.S. Air Force.
    #
    Joint USAF/NOAA Solar Region Summary
    SRS Number 71 Issued at 0030Z on 12 Mar 2001
    Report compiled from data received at SWO on 11 Mar
    I.  Regions with Sunspots.  Locations Valid at 11/2400Z
    Nmbr Location  Lo  Area  Z   LL   NN Mag Type
    9368 N25W73   089  0210 Cao  12   07 Beta
    9370 N10W51   067  0150 Dao  10   20 Beta
    9372 S36W76   092  0110 Dao  10   07 Beta
    9373 S08E48   328  0230 Dai  10   19 Beta
    9374 S18E38   338  0000 Axx  01   01 Alpha
    9375 S16W58   074  0010 Bxo  03   03 Beta
    9376 S13E72   304  0030 Hrx  01   01 Alpha
    9377 S18E01   015  0000 Axx  00   01 Alpha
    IA. H-alpha Plages without Spots.  Locations Valid at 11/2400Z Mar
    Nmbr  Location  Lo
    9366  S24W74   090
    9367  N09W71   086
    II. Regions Due to Return 12 Mar to 14 Mar
    Nmbr Lat    Lo
    9354 S09    285
    9357 S25    256
    9358 S10    248
    99999
    PLAIN
    Corrected Rgn 9377 latitude to S18
    ```

### Current Problem 

```python
from sunpy.io.special import srs
from sunpy.net import Fido
from sunpy.net import attrs as a

result = Fido.search(
    a.Time(
        datetime.datetime(2000, 1, 1), 
        datetime.datetime(2000, 1, 5),
        ),
    a.Instrument.soon,
)

table = Fido.fetch(result)

for filepath in table:
  try:
    srs_table = srs.read_srs(filepath)
  except Exception as e:
    print(f"filepath: {filepath} {str(e)[0:65]}\n") # truncate string
```
```shell
filepath: /root/sunpy/data/20000101SRS.txt 
ERROR: Unable to guess table format with the guesses listed belo

filepath: /root/sunpy/data/20000102SRS.txt 
ERROR: Unable to guess table format with the guesses listed belo

filepath: /root/sunpy/data/20000103SRS.txt 
ERROR: Unable to guess table format with the guesses listed belo

filepath: /root/sunpy/data/20000104SRS.txt 
ERROR: Unable to guess table format with the guesses listed belo

filepath: /root/sunpy/data/20000105SRS.txt 
ERROR: Unable to guess table format with the guesses listed belo
```

### Solution Output

```shell
 ID Number Carrington Longitude  Area  Z  Longitudinal Extent Number of Sunspots Mag Type Latitude Longitude
                   deg           uSH              deg                                       deg       deg   
--- ------ -------------------- ----- --- ------------------- ------------------ -------- -------- ---------
  I   8809                 27.0   0.0 BXO                 6.0                  2     BETA     24.0      20.0
  I   8810                 12.0 330.0 FAO                16.0                 16     BETA     38.0       5.0
  I   8812                 54.0   0.0 AXX                 0.0                  1    ALPHA    -27.0      47.0
  I   8813                329.0  40.0 DRO                 8.0                  8     BETA    -17.0     -38.0
  I   8814                320.0 130.0 HAX                 3.0                  2    ALPHA     10.0     -47.0
  I   8815                306.0  30.0 CRO                 5.0                  2     BETA     10.0     -61.0
 IA   8811                 96.0   ———  --                 ———                 --       --    -30.0      89.0
 II   8796                250.0   ———  --                 ———                 --       --    -25.0       ——— 

 ID Number Carrington Longitude  Area  Z  Longitudinal Extent Number of Sunspots Mag Type Latitude Longitude
                   deg           uSH              deg                                       deg       deg   
--- ------ -------------------- ----- --- ------------------- ------------------ -------- -------- ---------
  I   8810                 12.0 270.0 FAO                16.0                 11     BETA     38.0      18.0
  I   8813                333.0  50.0 DSO                 7.0                  6     BETA    -17.0     -21.0
  I   8814                320.0 180.0 HHX                 3.0                  3    ALPHA     10.0     -34.0
  I   8815                306.0  40.0 CRO                 8.0                  9     BETA     10.0     -48.0
 IA   8809                 27.0   ———  --                 ———                 --       --     24.0      33.0
 IA   8812                 54.0   ———  --                 ———                 --       --    -27.0      60.0
 II   8796                250.0   ———  --                 ———                 --       --    -25.0       ———
 II   8797                235.0   ———  --                 ———                 --       --    -32.0       ——— 

 ID Number Carrington Longitude  Area  Z  Longitudinal Extent Number of Sunspots Mag Type Latitude Longitude
                   deg           uSH              deg                                       deg       deg   
--- ------ -------------------- ----- --- ------------------- ------------------ -------- -------- ---------
  I   8810                 11.0 180.0 EAO                15.0                  7     BETA     37.0      29.0
  I   8813                334.0  60.0 DAO                 8.0                  7     BETA    -16.0      -8.0
  I   8814                321.0 150.0 HAX                 3.0                  3    ALPHA     10.0     -21.0
  I   8815                307.0  70.0 CAO                 8.0                 12     BETA      8.0     -35.0
 IA   8809                 27.0   ———  --                 ———                 --       --     24.0      46.0
 IA   8812                 54.0   ———  --                 ———                 --       --    -27.0      73.0
 II   8796                250.0   ———  --                 ———                 --       --    -25.0       ———
 II   8797                235.0   ———  --                 ———                 --       --    -32.0       ——— 

 ID Number Carrington Longitude  Area  Z  Longitudinal Extent Number of Sunspots Mag Type Latitude Longitude
                   deg           uSH              deg                                       deg       deg   
--- ------ -------------------- ----- --- ------------------- ------------------ -------- -------- ---------
  I   8810                 11.0 160.0 ESO                15.0                  4     BETA     38.0      42.0
  I   8813                336.0 130.0 DAO                 8.0                 14     BETA    -16.0       7.0
  I   8814                321.0 140.0 CAO                 4.0                  4     BETA     11.0      -8.0
  I   8815                307.0  50.0 CAO                 8.0                 15     BETA     10.0     -22.0
 IA   8809                 27.0   ———  --                 ———                 --       --     24.0      59.0
 IA   8812                 54.0   ———  --                 ———                 --       --    -27.0      86.0
 II   8797                235.0   ———  --                 ———                 --       --    -32.0       ——— 

 ID Number Carrington Longitude  Area  Z  Longitudinal Extent Number of Sunspots Mag Type Latitude Longitude
                   deg           uSH              deg                                       deg       deg   
--- ------ -------------------- ----- --- ------------------- ------------------ -------- -------- ---------
  I   8810                  7.0 170.0 FSO                16.0                  3     BETA     38.0      51.0
  I   8813                335.0 140.0 DSO                 8.0                 12     BETA    -16.0      20.0
  I   8814                321.0 100.0 HAX                 3.0                  2    ALPHA     12.0       6.0
  I   8815                304.0  20.0 BXO                 5.0                 14     BETA     10.0     -11.0
  I   8816                330.0  20.0 CRO                 3.0                  6     BETA     23.0      15.0
  I   8817                305.0  10.0 CRO                 4.0                  5     BETA     24.0     -10.0
 IA   8809                 27.0   ———  --                 ———                 --       --     24.0      72.0 
 ```

CC @alasdairwilson @samaloney

### CODE REVIEW

**Comment by PaulJWright** [comment]
**This is to be squashed before merge.** 

There are also some outstanding questions:
 - [ ] Do we want to cast McIntosh/Hale classifications to titlecase?
 - [ ] Add sections after IA to metadata?

File: changelog/7035.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Modifies ``sunpy.io.special.srs`` to correctly handle uppercase SRS files and supplementary sections occurring after the main data sections (I., IA, II). In addition to these changes, the new function ``sunpy.io.special.srs._try_drop_column()`` is used to remove the frequently empty `COMMENT` column in seen in section `IA`.
```
Comment (samaloney) [inline changelog/7035.bugfix.rst:1]:
```suggestion
Modifies ``sunpy.io.special.srs`` to correctly handle uppercase SRS files and supplementary sections occurring after the main data sections (I., IA, II). In addition to these changes, the new function ``sunpy.io.special.srs._try_drop_column()`` is used to remove the frequently empty `COMMENT` column as seen in section `IA`.
```

File: sunpy/data/test/19960106SRS.txt
Comment (samaloney) [inline sunpy/data/test/19960106SRS.txt:1]: maybe rename so it's more clear I guess this is, I guess this is an all caps example so maybe `all_caps_srs.txt` but I see the other files have original names so maybe not ...

**Review by samaloney — APPROVED** [review]
Looks good to me

**Comment by samaloney** [comment]
The circle CI test fail is unrelated I thinks as it's caused by xarray dip and the code coverage tests failed because the tests didn't actually run seem the GitHub action workflow needs to be approved to run 🤷 

> Do we want to cast McIntosh/Hale classifications to titlecase?

I would like this but I'm not sure it is the responsibly of sunpy to alter/fix the data as reported in the SRS files. 

> Add sections after IA to metadata?
Adding this would be great makes a more complete record

File: sunpy/data/test/19960106SRS.txt
Comment (PaulJWright) [reply sunpy/data/test/19960106SRS.txt:1]: Yeah, I think leaving it as is (there are a bunch of all caps SRS with various random sections)

**Comment by alasdairwilson** [comment]
Definitely leave the McIntosh case as is and agree that it would be nice to parse as much of the file as possible.

**Comment by samaloney** [comment]
I think need to change the quotes/backticks around `COMMENT` and `IA` some where  https://github.com/sunpy/sunpy/actions/runs/5119015219/jobs/9203901046?pr=7035#step:10:335

**Review by alasdairwilson — COMMENTED** [review]
Looks good. 👍

File: changelog/7035.bugfix.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Modifies ``sunpy.io.special.srs`` to correctly handle uppercase SRS files and supplementary sections occurring after the main data sections (I., IA, II). In addition to these changes, the new function ``sunpy.io.special.srs._try_drop_column()`` is used to remove the frequently empty `COMMENT` column as seen in section `IA`.
```
[the remark this answers is not in the document]
Comment (PaulJWright) [reply changelog/7035.bugfix.rst]: Let's see if that fixes the docs issue

**Review by alasdairwilson — APPROVED** [review]

File: changelog/7035.bugfix.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (alasdairwilson) [reply changelog/7035.bugfix.rst]: I only noticed the comment about them failing after removing them for other reasons but yeah, double backticks to look like `this` single backticks is sphinx refs.

File: sunpy/io/tests/test_srs.py [hunk 3]
```diff
@@ -55,3 +62,113 @@ def test_parse_location(loc_column, exp_longitude, exp_latitude):
...
+    filtered_dict = {key: value for key, value in expected_pattern_dict.items() if key in keys_to_keep}
+    result = srs._try_drop_empty_column(column_name_to_drop, data_lines, filtered_dict)
+    out = data_lines.copy()
+    out[0] = ' '.join([col.title() for col in out[0].split()][:-1])
+    assert result == out
+
+def test_tdec_too_small_dict_example(data_lines, expected_pattern_dict, column_name_to_drop):
+    """
+    columns aren't a subset of the dictionary
+    (no pattern to match for column `LO`)
```
Comment (nabobalis) [inline sunpy/io/tests/test_srs.py:106]:
```suggestion
    Columns aren't a subset of the dictionary: no pattern to match for column `LO`
```

**Review by nabobalis — APPROVED** [review]
Really minor comments.

File: changelog/7035.bugfix.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Modified :func:`sunpy.io.special.srs_read_srs` to correctly handle uppercase SRS files and supplementary sections occurring after the main data sections (I., IA, II).
```
Comment (samaloney) [inline changelog/7035.bugfix.rst:1]:
```suggestion
Modified :func:`sunpy.io.special.read_srs` to correctly handle uppercase SRS files and supplementary sections occurring after the main data sections (I., IA, II).
```

Comment (PaulJWright) [reply changelog/7035.bugfix.rst]: @alasdairwilson ;)

File: sunpy/io/special/srs.py [hunk 5]
```diff
@@ -228,3 +255,77 @@ def parse_lat_col(column, latitude_column):
...
+
+    """
+    if not isinstance(column_name_to_drop, str):
+        raise ValueError("``column_name_to_drop`` must be a string.")
+
+    if not isinstance(data_lines, list) or len(data_lines) < 2:
+        raise ValueError("``data_lines`` must be a list with at least two elements.")
+
+    if not isinstance(pattern_dict, dict):
+        raise ValueError("``pattern_dict`` must be a dictionary.")
```
[the remark this answers is not in the document]
Comment (PaulJWright) [reply sunpy/io/special/srs.py:288]: Happy to remove them and the associated tests

File: changelog/7035.bugfix.rst (hunk 4 above)
Comment (alasdairwilson) [reply changelog/7035.bugfix.rst]: oops

File: changelog/7035.bugfix.rst [hunk 6]
```diff
@@ -0,0 +1 @@
+Modified :func:`sunpy.io.special.read_srs` to correctly handle uppercase SRS files and supplementary sections occurring after the main data sections (I., IA, II).
```
Comment (nabobalis) [inline changelog/7035.bugfix.rst:1]:
```
<unknown>:7: WARNING: py:func reference target not found: sunpy.io.special.read_srs
```
Unsure what the problem is, it is imported here at this level.

from sunpy.io.special.srs import read_srs - Works
import sunpy.io.special.read_srs - Does not work.

**Comment by PaulJWright** [comment]
> The original post mentions now only 0.3% error.
> 
> What are those errors?

Oh, I added that a while back. Basically if you try to load every SRS from 1996 - 2023, 17% would fail going through `read_srs()`. Now ~0.3% fail (I think that's 27 or so of the 10,000). There are some empty files, some random `99999` on lines, some missing lat/long etc., and other minor one-off things that cause errors during the `srs_read()`.

**Comment by wtbarnes** [comment]
A few overarching comments:

- I don't think we can backport this because of the column renaming. I don't think this needs a deprecation period but potentially breaking code on a bug fix release is not ideal. Happy to hear other arguments though. 
- Because a bunch of new test data has been added, it'd be better to put this in an `srs` subdirectory in the test data folder, particularly since the file names are kind of opaque.

**Comment by alasdairwilson** [comment]
> A few overarching comments:
> 
> * I don't think we can backport this because of the column renaming. I don't think this needs a deprecation period but potentially breaking code on a bug fix release is not ideal. Happy to hear other arguments though.


I don't think this is correct, it is my understanding that newer SRS files were all already title case and these were the only one sunpy was able to parse. Older files are upper case and this casts them as title case so the column names are consistent with what was previously being output by `read_srs`. Perhaps @PaulJWright can clarify though.

As for moving the tests into srs subdirectory : 👍

File: changelog/7035.bugfix.rst (hunk 6 above)
Comment (alasdairwilson) [inline changelog/7035.bugfix.rst:1]:
```suggestion
Modified :func:`sunpy.io.special.srs.read_srs` to correctly handle uppercase SRS files and supplementary sections occurring after the main data sections (I., IA, II).
```

**Comment by PaulJWright** [comment]
> > A few overarching comments:
> > 
> > * I don't think we can backport this because of the column renaming. I don't think this needs a deprecation period but potentially breaking code on a bug fix release is not ideal. Happy to hear other arguments though.
> 
> I don't think this is correct, it is my understanding that newer SRS files were all already title case and these were the only one sunpy was able to parse. Older files are upper case and this casts them as title case so the column names are consistent with what was previously being output by `read_srs`. Perhaps @PaulJWright can clarify though.
> 
> As for moving the tests into srs subdirectory : 👍

yes, the current SRS files are "titlecase" by default. This change should only affect the uppercase SRS files that couldn't be loaded anyway.

**Comment by PaulJWright** [comment]
ok @alasdairwilson @wtbarnes, files should have been moved correctly.

**Comment by nabobalis** [comment]
I will manually backport this.

**Pull request merged**